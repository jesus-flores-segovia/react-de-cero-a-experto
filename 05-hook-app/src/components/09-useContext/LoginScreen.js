import React, { useContext } from 'react';
import { useForm } from '../../hooks/useForm';
import { UserContext } from './UserContext';

export const LoginScreen = () => {

    const [values, handleInputChange, reset] = useForm({
        email: "",
        password: "",
    });

    const {email, password} = values;

    const {user, setUser} = useContext(UserContext);

    const handleSubmit = (e) => {
        e.preventDefault();
        setUser(values);
        reset();
        console.log(values);
    };

    return (
        <div>
            <h1>LoginScreen</h1>
            <hr/>
            <form onSubmit={handleSubmit}>
                <div className="mb-3">
                    <label htmlFor="inputEmail" className="form-label">Email address</label>
                    <input name="email" type="email" className="form-control" id="inputEmail" aria-describedby="emailHelp" value={email} onChange={handleInputChange}/>
                </div>
                <div className="mb-3">
                    <label htmlFor="inputPassword" className="form-label">Password</label>
                    <input name="password" type="password" className="form-control" id="inputPassword" value={password} onChange={handleInputChange}/>
                </div>
                <button type="submit" className="btn btn-primary">Login</button>
            </form>
        </div>
    )
}
