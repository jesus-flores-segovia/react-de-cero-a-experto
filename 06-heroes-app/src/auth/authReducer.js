import { types } from "../types/types";

export const authReducer = (state = {}, action) => {

    switch (action.type) {
        case types.login:
            return {
                ...state,
                name: action.payload,
                logged: true
            }
        case types.logout:
            return {
                ...state,
                name: "",
                logged: false
            }
    
        default:
            return state;
    }
}