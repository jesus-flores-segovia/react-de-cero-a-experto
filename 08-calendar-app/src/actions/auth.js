import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarLogoutCleanEvents } from "./calendar";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth",
        { email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { email, password } = body.errors;
          const message = email ? email.msg : password.msg;
          Swal.fire("Cannot login", message, "error");
          return;
        }

        Swal.fire("Cannot login", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth/register",
        { name, email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { name, email, password } = body.errors;
          const message = name ? name.msg : email ? email.msg : password.msg;
          Swal.fire("Cannot register", message, "error");
          return;
        }

        Swal.fire("Cannot register", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startCheckingLoginState = () => {
  return async (dispatch) => {
    try {
      dispatch(checkLoginState({ isCheckingLoginState: true }));
      const response = await fetchWithToken("auth/newToken");
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      }
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    } catch (error) {
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarLogoutCleanEvents());
    dispatch(logout());
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkLoginState = (isCheckingLoginState) => ({
  type: types.authCheckLoginState,
  payload: isCheckingLoginState,
});

const logout = () => ({
  type: types.authLogout,
});
