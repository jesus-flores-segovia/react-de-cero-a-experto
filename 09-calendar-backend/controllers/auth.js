const { response } = require("express");
const { validationResult } = require("express-validator");

const register = (req, res = response) => {
  const { name, email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "register", name, email, password });
};

const login = (req, res = response) => {
  const { email, password } = req.body;
  const errors = validationResult(req);

  if (!errors.isEmpty()) {
    return res.status(400).send({ ok: false, errors: errors.mapped() });
  }

  res.status(201).send({ ok: true, msg: "login", email, password });
};

const newToken = (req, res = response) => {
  res.send({ ok: true, msg: "new token" });
};

module.exports = {
  register,
  login,
  newToken,
};
