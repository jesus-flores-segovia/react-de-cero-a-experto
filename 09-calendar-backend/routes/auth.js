/* 
    Routing for the users auth
    url: host + /api/auth
*/

const express = require("express");
const { check } = require("express-validator");
const router = express.Router();

const { register, login, newToken } = require("../controllers/auth");

router.post(
  "/register",
  [
    check("name", "Name is mandatory").not().isEmpty(),
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  register
);

router.post(
  "/",
  [
    check("email", "Email is mandatory").isEmail(),
    check("password", "Password must be at least of 6 characters").isLength({
      min: 6,
    }),
  ],
  login
);

router.get("/newToken", newToken);

module.exports = router;
