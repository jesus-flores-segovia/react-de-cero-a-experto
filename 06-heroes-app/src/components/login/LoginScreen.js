import React, { useContext } from 'react';
import { login } from '../../actions/authActions';
import { AuthContext } from '../../auth/AuthContext';

export const LoginScreen = ({history}) => {

    const {dispatch} = useContext(AuthContext);

    const handleLogin = () => {
        dispatch(login("Jes√∫s"));
        history.replace("/");
    }

    return (
        <div className="container mt-5">
            <h1>Login</h1>
            <hr/>
            <button 
                className="btn btn-primary"
                onClick={handleLogin}>
                Login
            </button>
        </div>
    )
}
