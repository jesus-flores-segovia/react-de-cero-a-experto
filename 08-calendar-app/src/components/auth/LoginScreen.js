import React from "react";
import { useDispatch } from "react-redux";
import Swal from "sweetalert2";
import { startLogin, startRegister } from "../../actions/auth";
import { isValidEmail } from "../../helpers/validator";
import { useForm } from "../../hooks/useForm";
import "./login.css";

export const LoginScreen = () => {
  const dispatch = useDispatch();

  const [formLoginValues, handleLoginInputChange] = useForm({
    loginEmail: "",
    loginPassword: "",
  });

  const { loginEmail, loginPassword } = formLoginValues;

  const handleLogin = (e) => {
    e.preventDefault();

    if (loginEmail.trim().length === 0) {
      Swal.fire("Cannot login", "Email is mandatory", "error");
      return;
    } else if (!isValidEmail(loginEmail)) {
      Swal.fire("Cannot login", "Email is not valid", "error");
      return;
    }

    if (!loginPassword.length >= 6) {
      Swal.fire(
        "Cannot login",
        "Password must be at least of 6 characters",
        "error"
      );
    }

    dispatch(startLogin(loginEmail, loginPassword));
  };

  const [formRegisterValues, handleRegisterInputChange] = useForm({
    registerName: "",
    registerEmail: "",
    registerPassword: "",
    registerPassword2: "",
  });

  const { registerName, registerEmail, registerPassword, registerPassword2 } =
    formRegisterValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if (registerName.trim().length === 0) {
      Swal.fire("Cannot register", "Name is mandatory", "error");
      return;
    }

    if (registerEmail.trim().length === 0) {
      Swal.fire("Cannot register", "Email is mandatory", "error");
      return;
    } else if (!isValidEmail(registerEmail)) {
      Swal.fire("Cannot register", "Email is not valid", "error");
      return;
    }

    if (registerPassword.length < 6) {
      Swal.fire(
        "Cannot register",
        "Password must be at least of 6 characters",
        "error"
      );
      return;
    }

    if (registerPassword !== registerPassword2) {
      Swal.fire("Cannot register", "Password fields must match", "error");
    }

    dispatch(startRegister(registerName, registerEmail, registerPassword));
  };

  return (
    <div className="container login-container">
      <div className="row">
        <div className="col-md-6 login-form-1">
          <h3>Login</h3>
          <form onSubmit={handleLogin} autoComplete="off">
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Email"
                name="loginEmail"
                value={loginEmail}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name="loginPassword"
                value={loginPassword}
                onChange={handleLoginInputChange}
              />
            </div>
            <div className="form-group">
              <input type="submit" className="btnSubmit" value="Login" />
            </div>
          </form>
        </div>

        <div className="col-md-6 login-form-2">
          <h3>Register</h3>
          <form onSubmit={handleRegister}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Name"
                name="registerName"
                value={registerName}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Email"
                name="registerEmail"
                value={registerEmail}
                onChange={handleRegisterInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Password"
                name="registerPassword"
                value={registerPassword}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repeat the password"
                name="registerPassword2"
                value={registerPassword2}
                onChange={handleRegisterInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="submit"
                className="btnSubmit"
                value="Create account"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
