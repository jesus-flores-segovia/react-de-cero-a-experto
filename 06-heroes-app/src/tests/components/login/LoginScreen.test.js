import { mount } from "enzyme"
import { AuthContext } from "../../../auth/AuthContext";
import { LoginScreen } from "../../../components/login/LoginScreen"

describe("Tests inside file 'LoginScreen.test.js'", () => {

    const history = {
        replace: jest.fn(),
    };

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: true,
            name:"Jesús"
        }
    }

    beforeEach(() => {
        jest.clearAllMocks()
    });

    test("LoginScreen component must be shown correctly", () => {

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <LoginScreen history={history}/>
            </AuthContext.Provider>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("h1").text().trim()).toBe("Login")
    })

    test("LoginScreen component must log the user when the login button is clicked", () => {

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <LoginScreen history={history}/>
            </AuthContext.Provider>
        );

        wrapper.find("button").prop("onClick")();
        expect(contextValue.dispatch).toHaveBeenCalledTimes(1);
        expect(contextValue.dispatch).toHaveBeenCalledWith({payload: "Jesús", type: "[auth] login"})
        expect(history.replace).toHaveBeenCalledTimes(1);
        expect(history.replace).toHaveBeenCalledWith('/');
    })

    test("LoginScreen component must retrieve the last path saved in the local storage", () => {
        
        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <LoginScreen history={history}/>
            </AuthContext.Provider>
        );

        localStorage.setItem("lastPath", "/dc");

        wrapper.find("button").prop("onClick")();
        expect(history.replace).toHaveBeenCalledTimes(1);
        expect(history.replace).toHaveBeenCalledWith('/dc');
    })
})