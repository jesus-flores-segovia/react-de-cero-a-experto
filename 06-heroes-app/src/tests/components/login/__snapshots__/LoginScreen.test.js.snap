// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'LoginScreen.test.js' LoginScreen component must be shown correctly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
